### PR TITLE
Fix term update with same slugs

### PR DESCRIPTION
**Pull request by Hug0-Drelon**
Fixes https://github.com/polylang/polylang-pro/issues/1468

## Changes
- Use `term_exists_by_slug()` directly in `PLL_CRUD_Terms::set_pre_term_slug()`.
- Do not suffix with language if we are editing the term (i.e. `$_POST['tag_id']` is set and equal to result of `term_exists_by_slug()`)
- Handle post parent with same slug by prepending it to itself like WordPress does.
- Add some tests.

### CODE REVIEW

File: include/crud-terms.php [hunk 1]
```diff
@@ -307,25 +276,52 @@ private function get_slug_separator() {
...
+		 * @param string            $taxonomy Term taonomy.
 		 * @param string            $slug     Term slug
-		 * @param string            $taxonomy Term taxonomy.
 		 */
-		$lang = apply_filters( 'pll_inserted_term_language', null, $name, $taxonomy );
+		$lang = apply_filters( 'pll_inserted_term_language', null, $taxonomy, $slug );
 
-		if ( $lang instanceof PLL_Language && ! $this->model->term_exists_by_slug( $name, $lang, $taxonomy ) ) {
-			$slug = $name . $this->get_slug_separator() . $lang->slug;
+		if ( $lang instanceof PLL_Language ) {
```
Comment (Screenfeed) [inline include/crud-terms.php:298]: Here too, early return please.

File: include/crud-terms.php [hunk 2]
```diff
@@ -307,25 +276,52 @@ private function get_slug_separator() {
...
+			$parent = apply_filters( 'pll_inserted_term_parent', 0, $taxonomy, $slug );
+
+			/**
+			 * Special case if the parent has the same slug.
+			 * WordPress prepends the slug to itself, so let's do it.
+			 */
+			$parent_obj = get_term( $parent, $taxonomy );
+			if ( $parent_obj instanceof WP_Term && $parent_obj->slug === $slug ) {
+				return $slug .= '-' . $slug;
+			}
```
Comment (Screenfeed) [inline include/crud-terms.php:317]: Same comment I made [here](https://github.com/polylang/polylang-pro/pull/1490/files#r993631909).

File: tests/phpunit/tests/test-admin-filters-term.php [hunk 3]
```diff
@@ -730,4 +730,38 @@ public function test_update_term_from_admin() {
...
+		$new_slug      = 'well-tested';
+
+		$cat_en = $this->factory()->category->create_and_get(
+			array(
+				'name' => $original_name,
+			)
+		);
+		self::$model->term->set_language( $cat_en->term_id, 'en' );
+
+		$this->assertSame( sanitize_title( $original_name ), $cat_en->slug, 'The category slug is well created.' );
```
Comment (Screenfeed) [inline tests/phpunit/tests/test-admin-filters-term.php:745]: I'm not sure about this error message x) (if it's "well created", then what is the problem?).
#CopyPasteDrama

File: tests/phpunit/tests/test-admin-filters-term.php [hunk 4]
```diff
@@ -730,4 +730,38 @@ public function test_update_term_from_admin() {
...
+
+		// Add globals like an admin request.
+		$_REQUEST = $_POST = array(
+			'term_lang_choice' => 'en',
+			'_pll_nonce'       => wp_create_nonce( 'pll_language' ),
+		);
+
+		// Now update the category with a new term.
```
Comment (Screenfeed) [inline tests/phpunit/tests/test-admin-filters-term.php:753]:
```suggestion
		// Now update the category with a new slug.
```

File: tests/phpunit/tests/test-admin-filters-term.php [hunk 5]
```diff
@@ -730,4 +730,38 @@ public function test_update_term_from_admin() {
...
+		$updated_cat = wp_update_term(
+			$cat_en->term_id,
+			'category',
+			array(
+				'slug' => $new_slug,
+			)
+		);
+		$updated_cat_obj = get_term( $updated_cat['term_id'] );
```
Comment (Screenfeed) [inline tests/phpunit/tests/test-admin-filters-term.php:761]:
```suggestion
		$this->assertIsArray( $updated_cat );
		$updated_cat_obj = get_term( $updated_cat['term_id'] );
```

File: tests/phpunit/tests/test-slugs.php [hunk 6]
```diff
@@ -12,20 +19,122 @@ public static function wpSetUpBeforeClass( WP_UnitTest_Factory $factory ) {
...
+		$this->assertInstanceOf( WP_Term::class, $en_parent );
+		$this->assertEquals( 'test', $en_parent->slug );
+
+		$_POST['term_lang_choice'] = 'en';
+		$_POST['parent']           = $en_parent->term_id;
+		$en                        = self::factory()->term->create_and_get( array( 'taxonomy' => 'category', 'name' => 'test', 'parent' => $en_parent->term_id ) );
+		$this->pll_admin->model->term->set_language( $en, 'en' );
+
+		$this->assertInstanceOf( WP_Term::class, $en );
+		$this->assertEquals( 'test-test', $en->slug );
```
Comment (Screenfeed) [inline tests/phpunit/tests/test-slugs.php:75]: This is exactly what `test_term_with_parents_sharing_same_name()` does, so I think `test_term_with_parents_sharing_same_name()` can be removed.

File: tests/phpunit/tests/test-slugs.php [hunk 7]
```diff
@@ -12,20 +19,122 @@ public static function wpSetUpBeforeClass( WP_UnitTest_Factory $factory ) {
...
+	}
+
+	public function test_translated_terms_with_parents_sharing_same_name() {
+		$en_parent = self::factory()->term->create_and_get( array( 'taxonomy' => 'category', 'name' => 'test' ) );
+		$this->pll_admin->model->term->set_language( $en_parent->term_id, 'en' );
+
+		$this->assertInstanceOf( WP_Term::class, $en_parent );
+		$this->assertEquals( 'test', $en_parent->slug );
```
Comment (Screenfeed) [inline tests/phpunit/tests/test-slugs.php:67]:
```suggestion
		$this->assertSame( 'test', $en_parent->slug );
```

image

**Review by Screenfeed — CHANGES_REQUESTED** [review]

File: include/crud-terms.php (hunk 2 above)
Comment (Hug0-Drelon) [reply include/crud-terms.php]: Good idea! Done in 6728fe0

File: include/crud-terms.php (hunk 1 above)
Comment (Hug0-Drelon) [reply include/crud-terms.php]: Done in 6728fe0

File: tests/phpunit/tests/test-slugs.php (hunk 7 above)
Comment (Hug0-Drelon) [reply tests/phpunit/tests/test-slugs.php:67]: Ouch! Copy/paste hell!

Comment (Hug0-Drelon) [reply tests/phpunit/tests/test-slugs.php]: Done in a0999f0

File: tests/phpunit/tests/test-admin-filters-term.php (hunk 3 above)
Comment (Hug0-Drelon) [reply tests/phpunit/tests/test-admin-filters-term.php]: Yup! Done in 71acd76

File: admin/admin-filters-term.php [hunk 8]
```diff
@@ -654,19 +653,30 @@ public function get_inserted_term_language( $lang, $slug, $taxonomy ) {
...
+	 * Filters the subsequently inserted term parent in admin.
+	 *
+	 * @since 3.3
+	 *
+	 * @param int    $parent   Parent term ID, 0 if none found.
+	 * @param string $taxonomy Term taxonomy.
+	 * @return int Parent term ID if found, 0 otherwise.
+	 */
+	public function get_inserted_term_parent( $parent, $taxonomy ) {
+		if ( ! $parent ) {
```
[the remark this answers is not in the document]
Comment (Hug0-Drelon) [reply admin/admin-filters-term.php]: 👍 Done in 85b6745

**Review by Chouby — CHANGES_REQUESTED** [review]

**Review by Screenfeed — APPROVED** [review]

File: include/crud-terms.php [hunk 9]
```diff
@@ -307,25 +276,68 @@ private function get_slug_separator() {
...
+
+			return $slug .= $parent_suffix;
+		}
+
+		$term_id = (int) $this->model->term_exists_by_slug( $slug, $lang, $taxonomy, $parent );
 
-		if ( $lang instanceof PLL_Language && ! $this->model->term_exists_by_slug( $name, $lang, $taxonomy ) ) {
-			$slug = $name . $this->get_slug_separator() . $lang->slug;
+		// If no term exists in the given language with that slug, it can be created.
```
[the remark this answers is not in the document]
Comment (Hug0-Drelon) [reply include/crud-terms.php:338]: English class won't hurt ...

**Review by Screenfeed — APPROVED** [review]

**Pull request merged**